src/layout.cpp

```diff
@@ -0,0 +1,476 @@
+  degReinit(nodes);
+
+  std::vector<bool> usedId = {};
+  usedId.resize(nodes.size(), false);
+
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].degM == 0) {
+      bool exist = algorithmDFS(nodes[i].id, nodes, usedId);
+      if (exist) {
+        return true;
+      }
+    }
+  }
+
+  return algorithmDFS(nodes[0].id, nodes, usedId);
+}
+
+struct EdgeCount {
+  int degP = 0;
+  int degM = 0;
+};
+
+void removeEdgeCount(int i, std::vector<EdgeCount> &edgeCounts,
+                     std::vector<TreeNode> &nodes) {
+  for (int j = 0; j < nodes[i].succ.size(); j++) {
+    edgeCounts[nodes[i].succ[j]].degM--;
+  }
+  for (int j = 0; j < nodes[i].pred.size(); j++) {
+    edgeCounts[nodes[i].pred[j]].degP--;
+  }
+
+  edgeCounts[i].degM = -1;
+  edgeCounts[i].degP = -1;
+}
+
+void greedyFAS(
+    std::vector<TreeNode> &nodes,
+    std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> &deletedEdges) {
+  degReinit(nodes);
+  predReinit(nodes);
+  std::vector<TreeNode::nodeId> s1 = {}, s2 = {};
+  std::vector<EdgeCount> edgeCounts = {};
+  for (int i = 0; i < nodes.size(); i++) {
+    EdgeCount edgeCount = {};
+    edgeCount.degP = nodes[i].degP;
+    edgeCount.degM = nodes[i].degM;
+    edgeCounts.push_back(edgeCount);
+  }
+  int lenNodes = static_cast<int>(nodes.size());
+  while (lenNodes > 0) {
+    for (int i = 0; i < edgeCounts.size(); i++) {
+      if (edgeCounts[i].degP == 0) {
+        s2.insert(s2.cbegin(), i);
+        removeEdgeCount(i, edgeCounts, nodes);
+        lenNodes--;
+
+        i = -1;
+      }
+    }
+
+    for (int i = 0; i < edgeCounts.size(); i++) {
+      if (edgeCounts[i].degM == 0) {
+        s1.push_back(i);
+        removeEdgeCount(i, edgeCounts, nodes);
+        lenNodes--;
+
+        i = -1;
+      }
+    }
+
+    if (lenNodes > 0) {
+      int max;
+      for (int i = 0; i < edgeCounts.size(); i++) {
+        if (edgeCounts[i].degP >= 0 && edgeCounts[i].degM >= 0) {
+          max = i;
+          break;
+        }
+      }
+      for (int i = 0; i < edgeCounts.size(); i++) {
+        if (edgeCounts[i].degP - edgeCounts[i].degM > max &&
+            edgeCounts[i].degP >= 0 && edgeCounts[i].degM >= 0) {
+          max = i;
+        }
+      }
+
+      s1.push_back(max);
+      removeEdgeCount(max, edgeCounts, nodes);
+
+      lenNodes--;
+    }
+  }
+
+  lenNodes = static_cast<int>(nodes.size());
+  for (int i = 0; i < s1.size(); i++) {
+    nodes.push_back(nodes[s1[i]]);
+  }
+  for (int i = 0; i < s2.size(); i++) {
+    nodes.push_back(nodes[s2[i]]);
+  }
+  for (int i = 0; i < lenNodes; i++) {
+    nodes.erase(nodes.cbegin());
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      for (int k = 0; k < i; k++) {
+        if (nodes[i].succ[j] == nodes[k].id) {
+          auto edge = std::make_pair(nodes[i].id, nodes[i].succ[j]);
+          deletedEdges.push_back(edge);
+
+          nodes[i].succ.erase(nodes[i].succ.begin() + j);
```

Comment:
If an item is erased, `j` points to an unobserved item that will be skipped on the next iteration after `j++`. Consider reverse iteration order.